FixFor#40209 - Conditional TotalCollect Trigger for mutation and Virtual Carts

senthilengg opened the pull request:
### Description (*)

-  Trigger CollectTotals only if the below condition are true
-  Quote is not virtual
-  Check GraphQl ResolverInfo OperationType is not "query"

### Related Pull Requests
https://github.com/magento-gl/magento2ee/pull/713
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)

1. Fixes magento/magento2#40209

### Manual testing scenarios (*)
1. Include CartPrice in Cart Graphql
2. Also test all cart mutations

### Questions or comments
1. Similar GraphQl context check can be used in other areas to make the api's faster

Commit:
#40209 - Conditional TotalCollect Trigger for mutation and Virtual Carts

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

Commit:
#40209 - Fixing Implicit Null of ScopeConfig

Commit:
#40209 - Fixing Static Tests

Commit:
#40209 - Removing Extenstion Attribute after Shipping getData()

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Fix copyright comment formatting in CartPrices.php

senthilengg committed:
Fix copyright notice formatting in CartPricesTest.php

senthilengg reviewed (commented):
Resolved conflicts and Copyright format updated.

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Hotel committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

Commit:
#40209 - Fixing discount lable null and WebAPI Test Failures

senthilengg commented:
@magento run WebAPI Tests

Commit:
#40209 - Adding Discounts to graphql response retriving from extension_attributes

senthilengg commented:
@magento run WebAPI Tests

senthilengg commented:
@magento run Unit Tests

Commit:
#40209 - Made new params in constructor as optional and updated unit tests

Commit:
#40209 - Removed Extension Attributes and skip optimisation if discounts field is present in graphql request

Commit:
40209 - Removed Extension Attributes and skip optimisation if discounts field is present in graphql request

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

Commit:
#40209 - Removed unused variable

senthilengg commented:
@engcom-Hotel  Can you review this, I've fixed everything except the graphql error in B2B & EE which seems to be not related to this ticket.

senthilengg commented:
@magento run Static Tests

senthilengg commented:
@magento run all tests

Commit:
#40209 - Adding Trigger Recollect as a condition to fire collectTotals

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Remove unused TotalTestHelper import from CartPricesTest

Commit:
#40209 - Bringing back the helper removed from last merge

senthilengg commented:
@magento run Unit Tests, Static Tests, WebAPI Tests

engcom-Hotel committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/Resolver/CartPrices.php:
```diff
@@ -1,6 +1,6 @@
 <?php
 /**
- * Copyright 2019 Adobe
+ * Copyright 2025 Adobe
```
It should be 2019.

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/Resolver/CartPrices.php:
```diff
@@ -133,37 +191,59 @@ private function getAppliedTaxes(Total $total, string $currency): array
+     * @throws \InvalidArgumentException
      */
-    private function getSubtotalWithDiscountExcludingTax(Total $cartTotals): float
+    private function getSubtotalWithDiscountExcludingTax($totals): float
```
```suggestion
    private function getSubtotalWithDiscountExcludingTax(Total|CartTotals $totals): float
```

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/Resolver/CartPrices.php:
```diff
@@ -133,37 +191,59 @@ private function getAppliedTaxes(Total $total, string $currency): array
+     * @return void
+     * @throws \InvalidArgumentException If the provided totals instance is of an unsupported type.
+     */
+    private function validateTotalsInstance($totals)
```
Please specify the return type of this method.

engcom-Hotel reviewed (changes requested):
Thank you @senthilengg for the fixes. Please refer to the below review comments.

Commit:
#40209 - Addressing review comments

Commit:
40209 - Addressing review comments

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg commented:
@magento run Unit Tests, Static Tests, WebAPI Tests

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Hotel reviewed (changes requested):
Hello @senthilengg,

It seems the WebAPI tests are failing due to this PR changes. Please fix those.

Thank you

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

Commit:
Merge conflict resolved

Commit:
Merge conflict resolved

Commit:
40209-Merge conflict resolved

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Dash committed:
Fix WebAPI Failure

engcom-Dash committed:
Unit test failure fix

engcom-Dash commented:
@magento run all tests

senthilengg commented on app/code/Magento/QuoteGraphQl/Model/Resolver/CartPrices.php:
```diff
@@ -93,7 +93,8 @@ public function resolve(Field $field, $context, ResolveInfo $info, ?array $value
             $info->operation->operation == self::QUERY_TYPE &&
-            !array_key_exists('discounts', $info->getFieldSelection(1))
+            !array_key_exists('discounts', $info->getFieldSelection(1)) &&
+            !array_key_exists('gift_options', $info->getFieldSelection(1))
```
@engcom-Dash this will load collectTotals but it’s not required in a normal scenario. WebApi test failure is because of the script not setting address object. If it’s set then after add product to cart collectTotal will happen and update the GwPrice. So I would suggest to fix the script than introducing a condition here which is incorrect IMHO.

Adding a condition like `!array_key_exists('gift_options', $info->getFieldSelection(1))`
means, you’re modifying the feature’s behavior just to make the test pass — instead of ensuring the test setup correctly reflects the intended behavior or prerequisite of the feature. 

Similar condition was make sense for discounts since the fields inside discounts array is set as part of the colletTotals process. On top of this gift_options field itself from EE version which should not exists in open source.

engcom-Dash replied:
Thank you for your suggestion and feedback.
I’ve removed my condition changes from CartPrice.php and added the necessary changes in the test case file.

engcom-Dash committed:
Remove webapi failure changes from CartPrice.php

engcom-Dash committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Dash committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Dash committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

engcom-Dash commented:
@magento run all tests

engcom-Dash commented:
@magento run Unit Tests, Functional Tests EE, Functional Tests B2B, Functional Tests CE

engcom-Dash commented:
The consistent failures in Functional CE, B2B,EE are known issues and JIRA is open for the same. The other failures are inconsistent and flaky. They neither part of the PR nor failing because of the PR changes.

CE
Build 1
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40224/fb89a039f128d79d20584020975d49df/Functional/allure-report-ce/index.html#categories/8fb3a91ba5aaf9de24cc8a92edc82b5d
<img width="1240" height="463" alt="image" src="https://github.com/user-attachments/assets/e33dbec2-678a-4ae2-b971-7f31f6f03496" />

Build 2
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40224/2686a73cb88180ec32e045cfcaab89e0/Functional/allure-report-ce/index.html#categories/7c91e0ce9ab9dfaa9671d4f8bce7d3a9/cfa775b876566f16/
<img width="1260" height="289" alt="image" src="https://github.com/user-attachments/assets/311bcbfe-7599-40e2-b70d-39107340aeca" />


B2B
Build 1
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40224/bd80158a875192f5661f92ca784608e4/Functional/allure-report-b2b/index.html#categories/ed16d1362a7a61c31a1f41ef937defb9/6f838a50b1569a59/
<img width="1417" height="327" alt="image" src="https://github.com/user-attachments/assets/a0bd6324-cf96-402c-9127-04cbaf50ccd9" />

Build 2
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40224/b7fd8b1c174d533555713f7c7ecdf68f/Functional/allure-report-b2b/index.html#categories/ed16d1362a7a61c31a1f41ef937defb9/bb12e097d07cbd0a/
<img width="1418" height="364" alt="image" src="https://github.com/user-attachments/assets/f209fb1b-3e31-4269-96b5-f7b42be4e41a" />


EE
Build 1
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40224/f5af563716f578dcb9b22b696e693dda/Functional/allure-report-ee/index.html#categories/bdbf199525818fae7a8651db9eafe741
<img width="971" height="505" alt="image" src="https://github.com/user-attachments/assets/dbe20b48-307a-49e7-9cdd-08f7b677d80f" />

Build 2
https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/40224/81c7998a6e984e5b813b51de87f71745/Functional/allure-report-ee/index.html#categories
<img width="961" height="426" alt="image" src="https://github.com/user-attachments/assets/911515cf-0539-45f3-ae67-5a3d743b29e6" />

Known Issue
CheckProductsAfterUpdateDeleteStagingTest[ACQE-8667](https://jira.corp.adobe.com/browse/ACQE-8667)
AdminSearchCustomerAddressByKeywordTest[ACQE-8945](https://jira.corp.adobe.com/browse/ACQE-8945)

Hence, Moving this PR to Merge In Progress

engcom-Dash commented:
Hello @senthilengg,

Thanks for the contribution!

We are currently discussing this PR with our Product Owner to obtain the required approval. Hence, we are moving this PR on hold for now. We will keep you updated once the approval is received.

Thank you

engcom-Dash commented:
Hi @senthilengg,

Thank you for your response.

I have shared your inputs with the Product Owner. We will keep you updated once feedback is received from the PO.

Thank you

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
#40224 - Refactor totalsFactoryMock creation in CartPricesTest

Replaced the use of getMockBuilder with createPartialMockWithReflection for the totalsFactoryMock.

senthilengg committed:
#40209- Refactor totalsFactoryMock in CartPricesTest

Removed unused methods from totalsFactoryMock.

senthilengg committed:
#40209 - Add methods to quote mock in CartPricesTest

Commit:
#40209 - Fixed Static Test Issues

senthilengg committed:
#40209-Reverted copyright year in CartPricesTest.php

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance

senthilengg committed:
Merge branch '2.4-develop' into Fix-QuoteGraphQl-CartPrices-Performance
